File: myExpenses/src/test/java/org/totschnig/myexpenses/sync/SyncAdapterMergeChangeSetsTest.java
package org.totschnig.myexpenses.sync;

import org.junit.Test;
import org.totschnig.myexpenses.sync.json.TransactionChange;

import java.util.ArrayList;
import java.util.List;

import androidx.core.util.Pair;

import static org.junit.Assert.assertEquals;

public class SyncAdapterMergeChangeSetsTest extends SyncAdapterBaseTest {

  @Test
  public void noConflictsShouldBeReturnedIdentical() {
    List<TransactionChange> first = new ArrayList<>();
    first.add(buildCreated().setUuid("random1").build());
    List<TransactionChange> second = new ArrayList<>();
    second.add(buildCreated().setUuid("random2").build());
    Pair<List<TransactionChange>, List<TransactionChange>> result = syncDelegate.mergeChangeSets(first, second);
    assertEquals(first, result.first);
    assertEquals(second, result.second);
  }

  @Test
  public void deleteInSameSetShouldTriggerRemovalOfRelatedChanges() {
    String uuid = "random";
    List<TransactionChange> first = new ArrayList<>();
    first.add(buildCreated().setUuid(uuid).build());
    first.add(buildDeleted().setUuid(uuid).build());
    List<TransactionChange> second = new ArrayList<>();
    Pair<List<TransactionChange>, List<TransactionChange>> result = syncDelegate.mergeChangeSets(first, second);
    assertEquals(1, result.first.size());
    assertEquals(result.first.get(0),first.get(1));
    assertEquals(0, result.second.size());
  }

  @Test
  public void deleteInDifferentSetShouldTriggerRemovalOfRelatedChanges() {
    String uuid = "random";
    List<TransactionChange> first = new ArrayList<>();
    first.add(buildUpdated().setUuid(uuid).build());
    List<TransactionChange> second = new ArrayList<>();
    second.add(buildDeleted().setUuid(uuid).build());
    Pair<List<TransactionChange>, List<TransactionChange>> result = syncDelegate.mergeChangeSets(first, second);
    assertEquals(0, result.first.size());
    assertEquals(second, result.second);
  }

  @Test
  public void updatesShouldBeMerged() {
    String uuid = "random";
    List<TransactionChange> first = new ArrayList<>();
    first.add(buildUpdated().setUuid(uuid).build());
    first.add(buildUpdated().setUuid(uuid).build());
    List<TransactionChange> second = new ArrayList<>();
    Pair<List<TransactionChange>, List<TransactionChange>> result = syncDelegate.mergeChangeSets(first, second);
    assertEquals(1, result.first.size());
  }

  @Test
  public void insertCanBeMergedWithUpdate() {
    String uuid = "random";
    List<TransactionChange> first = new ArrayList<>();
    first.add(buildCreated().setUuid(uuid).build());
    first.add(buildUpdated().setUuid(uuid).build());
    List<TransactionChange> second = new ArrayList<>();
    Pair<List<TransactionChange>, List<TransactionChange>> result = syncDelegate.mergeChangeSets(first, second);
    assertEquals(1, result.first.size());
  }
}

File: myExpenses/src/test/java/org/totschnig/myexpenses/sync/SyncAdapterMergeChangeSetsTest.kt
package org.totschnig.myexpenses.sync

import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.totschnig.myexpenses.sync.json.TransactionChange

@RunWith(RobolectricTestRunner::class)
class SyncAdapterMergeChangeSetsTest : SyncAdapterBaseTest() {
    @Test
    fun listsWithoutMatchingShouldBeReturnedIdentical() {
        val first = listOf(
            buildCreated("random1")
                .copy(amount = 123)
        )
        val second = listOf(
            buildCreated("random2")
                .copy(comment = "Commentary")
        )
        val result = syncDelegate.mergeChangeSets(first, second)
        assertThat(result.first).isEqualTo(first)
        assertThat(result.second).isEqualTo(second)
    }

    @Test
    fun listsWithMatchingWithoutConflictShouldBeReturnedIdentical() {
        val first = listOf(
            buildCreated("random1")
                .copy(amount = 123)
        )
        val second = listOf(
            buildCreated("random1")
                .copy(comment = "Commentary")
        )
        val result = syncDelegate.mergeChangeSets(first, second)
        assertThat(result.first).isEqualTo(first)
        assertThat(result.second).isEqualTo(second)
    }

    @Test
    fun listsWithConflictShouldHonourLast() {
        val first = listOf(
            buildUpdated("random1").copy(
                timeStamp = 1,
                comment = "Kommentar"
            )
        )
        val second = listOf(
            buildUpdated("random1").copy(
                timeStamp = 2, comment = "Commentary"
            )
        )
        val result: Pair<List<TransactionChange>, List<TransactionChange>> =
            syncDelegate.mergeChangeSets(first, second)
        assertThat(result.first.first().comment).isNull()
        assertThat(result.second.first().comment).isEqualTo("Commentary")
    }

    @Test
    fun deleteInSameSetShouldTriggerRemovalOfRelatedChanges() {
        val uuid = "random"
        val deleteChange = buildDeleted(uuid)
        val first = listOf(
            buildCreated(uuid),
            deleteChange
        )
        val result = syncDelegate.mergeChangeSets(first, emptyList())
        assertThat(result.first).hasSize(1)
        assertThat(result.first.first()).isEqualTo(deleteChange)
        assertThat(result.second).isEmpty()
    }

    @Test
    fun cudWithSpecials() {
        val created = buildCreated("random1")
        val updated = buildUpdated("random1")
        val special = buildSpecial("random1")
        val first = listOf(
            created,
            updated,
            special
        )
        val result = syncDelegate.mergeChangeSets(first, emptyList())
        assertThat(result.first).isEqualTo(listOf(created, special))
    }

    @Test
    fun onlySpecials() {
        val special = buildSpecial("random1")
        val first = listOf(special)
        val result = syncDelegate.mergeChangeSets(first, emptyList())
        assertThat(result.first).isEqualTo(first)
    }

    @Test
    fun deleteInDifferentSetShouldTriggerRemovalOfRelatedChanges() {
        val uuid = "random"
        val first = listOf(buildUpdated(uuid))
        val second = listOf(buildDeleted(uuid))
        val result = syncDelegate.mergeChangeSets(first, second)
        assertThat(result.first).isEmpty()
        assertThat(result.second).isEqualTo(second)
    }

    @Test
    fun updatesShouldBeMerged() {
        val uuid = "random"
        val first = listOf(
            buildUpdated(uuid),
            buildUpdated(uuid)
        )
        val result = syncDelegate.mergeChangeSets(first, emptyList())
        assertThat(result.first).hasSize(1)
    }

    @Test
    fun insertCanBeMergedWithUpdate() {
        val uuid = "random"
        val first = listOf(
            buildCreated(uuid),
            buildUpdated(uuid)
        )
        val result = syncDelegate.mergeChangeSets(first, emptyList())
        assertThat(result.first).hasSize(1)
    }
}